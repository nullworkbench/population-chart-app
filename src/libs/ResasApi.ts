import axios, { AxiosError } from "axios";
import useSWR from "swr";

// RESAS-API独自エラーを出す用の型
export type RESASError = {
  statusCode: number;
  errorMessage: string;
};

// 都道府県の型
export type Prefecture = {
  prefCode: number;
  prefName: string;
};

// 総人口の型
export type Population = {
  prefCode: number;
  data: { year: number; value: number }[];
};

// resasApiを使うための基本設定を行なったaxiosインスタンス
export const resasApi = axios.create({
  baseURL: "https://opendata.resas-portal.go.jp/api/v1",
  headers: { "x-api-key": process.env.NEXT_PUBLIC_RESAS_API_KEY! },
});

// RESAS APIのエンドポイント
const endPoint = "https://opendata.resas-portal.go.jp/api/v1";

// RESAS APIキーをheaderに設定
const headers = { "X-API-KEY": process.env.NEXT_PUBLIC_RESAS_API_KEY! };

// 共通で利用するフェッチャー
const fetcher = (url: string) =>
  axios.get(url, { headers }).then((res) => res.data);

// 都道府県情報を取得する
export function usePrefectures() {
  const apiURL = "/api/prefectures";
  const fetcher2 = (url: string) => axios.get(url).then((res) => res.data);
  const { data, error } = useSWR(apiURL, fetcher2);

  return {
    // データは整形して返す
    prefectures: data as Prefecture[],
    isLoading: !error && !data,
    isError: error,
  };
}

// 指定されたprefCodeから総人口情報を取得する
export async function getPopulation(
  prefCode: number
): Promise<Population | void> {
  try {
    console.log("get");
    const res = await axios.get(
      `${endPoint}/population/composition/perYear?prefCode=${prefCode}&cityCode=-`,
      { headers }
    );
    // Resas独自のエラーをチェック
    const ResasError = isRESASError(res.data);
    if (ResasError) {
      console.log(`${ResasError.statusCode}: ${ResasError.errorMessage}`);
      return;
    }
    // 成功
    return { prefCode, data: res.data["result"]["data"][0]["data"] };
  } catch (error) {
    // エラー
    console.log(`Error getting population: ${error}`);
    return;
  }
}

// 取得は成功しているが、RESAS-APIのエラーがあるか判定する
export function isRESASError(data: any): RESASError | void {
  const objKeys = Object.keys(data);
  if (objKeys.includes("statusCode")) {
    // statusCodeがあるときは何らかのエラーが発生している
    return {
      statusCode: Number(data["statusCode"]),
      errorMessage: data["message"],
    };
  } else if (objKeys.length == 1 && objKeys[0] == "message") {
    // 429 Too Many Requests
    return { statusCode: 429, errorMessage: "Too Many Requests." };
  } else if (
    objKeys.length > 1 &&
    objKeys.includes("message") &&
    data[objKeys.indexOf("message")] == null
  ) {
    // 正しく取得できている
    console.log("getting data successfully");
    return;
  }
  // messageなしの何らかのエラー
  return { statusCode: 400, errorMessage: "Some Error Occured." };
}
