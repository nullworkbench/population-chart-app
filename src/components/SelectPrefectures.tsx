// 都道府県を選択するチェックボックス

import styled from "styled-components";

import { Prefecture } from "@/libs/ResasApi";
import { usePrefectures } from "@/libs/ResasApi";

import Checkbox from "@/components/ui/Checkbox";
import { ChangeEvent } from "react";

type Props = {
  // 親のuseStateを呼び出す
  selectedPrefs: Prefecture[];
  handleCheckboxChange: Function;
};

// 地方名
const regionNames = [
  { name: "北海道・東北", prefCodeRange: { min: 1, max: 7 } },
  { name: "関東", prefCodeRange: { min: 8, max: 14 } },
  { name: "北陸・信越", prefCodeRange: { min: 15, max: 20 } },
  { name: "東海", prefCodeRange: { min: 21, max: 24 } },
  { name: "近畿", prefCodeRange: { min: 25, max: 30 } },
  { name: "中国・四国", prefCodeRange: { min: 31, max: 39 } },
  { name: "九州・沖縄", prefCodeRange: { min: 40, max: 47 } },
];

const SelectPrefectures: React.FC<Props> = (prop) => {
  // 都道府県一覧を取得
  const { prefectures, isLoading, isError } = usePrefectures();

  if (isLoading) {
    return <p data-testid="loadingText">Loading...</p>;
  } else {
    const prefs = prefectures();
    if (prefs) {
      return (
        <div>
          {/* 地方ごとに分類して都道府県を表示 */}
          {regionNames.map((region, rIdx) => {
            return (
              <div key={rIdx}>
                <p>{region.name}</p>
                {prefs
                  .filter(
                    (p) =>
                      p.prefCode >= region.prefCodeRange.min &&
                      p.prefCode <= region.prefCodeRange.max
                  )
                  .map((pref, prefIdx) => (
                    <Checkbox
                      key={prefIdx}
                      checked={false}
                      label={pref.prefName}
                      handleOnChange={(e: ChangeEvent<HTMLInputElement>) =>
                        prop.handleCheckboxChange(e.target.checked, pref)
                      }
                    />
                  ))}
              </div>
            );
          })}
        </div>
      );
    } else {
      return <p data-testid="errorMessage">An Error Occured. Please Reload.</p>;
    }
  }
};

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export default SelectPrefectures;
