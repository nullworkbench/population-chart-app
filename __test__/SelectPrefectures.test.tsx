// テスト用ライブラリ
import {
  cleanup,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { rest } from "msw";
import { setupServer } from "msw/node";

// テスト対象
import SelectPrefectures from "@/components/SelectPrefectures";
import { SWRConfig } from "swr";
import { Prefecture } from "@/libs/ResasApi";

// APIのモックサーバーを立てる
const apiURL = "https://opendata.resas-portal.go.jp/api/v1/prefectures";
const handlers = [
  rest.get(
    // テスト中の以下のリクエストはこのモックサーバーに送られる
    apiURL,
    // ダミーデータ
    (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json({
          message: null,
          result: [
            {
              prefCode: 1,
              prefName: "北海道",
            },
            {
              prefCode: 2,
              prefName: "青森県",
            },
            {
              prefCode: 3,
              prefName: "岩手県",
            },
          ],
        })
      );
    }
  ),
];
const server = setupServer(...handlers);
beforeAll(() => {
  server.listen();
});
afterEach(() => {
  server.resetHandlers();
  cleanup();
});
afterAll(() => {
  server.close();
});

// テスト内容
describe("コンポーネントを描画", () => {
  // テスト１
  test("Loading...が表示される", async () => {
    // SWRのキャッシュを無効化して描画
    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SelectPrefectures selectedPrefs={[]} setSelectedPrefs={() => {}} />
      </SWRConfig>
    );
    expect(screen.getByTestId("loadingText"));
  });

  // テスト２
  test("47都道府県のチェックボックスが描画される", async () => {
    // SWRのキャッシュを無効化して描画
    const { asFragment } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SelectPrefectures selectedPrefs={[]} handleCheckboxChange={() => {}} />
      </SWRConfig>
    );

    // この要素が消えるまでテストを終了しないで待つ
    await waitForElementToBeRemoved(() =>
      // まずはSWRが取得中なのでLoadingが表示される
      expect(screen.getByTestId("loadingText"))
    );
    // 表示が完了すると、都道府県の情報が出る
    expect(screen.getByText("北海道"));
    expect(screen.getByText("青森県"));
    expect(screen.getByText("岩手県"));

    // 以前のテスト時のスナップショットと一致するか
    expect(asFragment()).toMatchSnapshot();
  });

  // テスト３
  test("APIからの取得時にエラーが発生した場合", async () => {
    // わざとエラーを発生させるためにモックサーバーを上書き
    server.use(
      rest.get(apiURL, (req, res, ctx) => {
        return res(ctx.status(403));
      })
    );
    // SWRのキャッシュを無効化して描画
    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SelectPrefectures selectedPrefs={[]} setSelectedPrefs={() => {}} />
      </SWRConfig>
    );
    // この要素が消えるまでテストを終了しないで待つ
    await waitForElementToBeRemoved(() =>
      // まずはSWRが取得中なのでLoadingが表示される
      expect(screen.getByTestId("loadingText"))
    );

    // エラー文が出ていることを確認
    expect(screen.getByTestId("errorMessage"));
  });
});
