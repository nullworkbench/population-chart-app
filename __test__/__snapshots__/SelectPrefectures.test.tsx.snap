// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`コンポーネントを描画 47都道府県のチェックボックスが描画される 1`] = `
<DocumentFragment>
  <div>
    <div
      class="SelectPrefectures__Wrapper-sc-qkgodw-0 byVAms"
    >
      <label
        class="Checkbox__Label-sc-1y2c1io-0 hLsZxk"
      >
        <input
          class="Checkbox__Input-sc-1y2c1io-1 dsaUmM"
          type="checkbox"
        />
        <span>
          北海道
        </span>
      </label>
      <label
        class="Checkbox__Label-sc-1y2c1io-0 hLsZxk"
      >
        <input
          class="Checkbox__Input-sc-1y2c1io-1 dsaUmM"
          type="checkbox"
        />
        <span>
          青森県
        </span>
      </label>
      <label
        class="Checkbox__Label-sc-1y2c1io-0 hLsZxk"
      >
        <input
          class="Checkbox__Input-sc-1y2c1io-1 dsaUmM"
          type="checkbox"
        />
        <span>
          岩手県
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
